### PR TITLE
[multistage][test] add with statement test

### PR DESCRIPTION
**Pull request by walterddr**
from PostgresQL doc [section 7.8.1](https://www.postgresql.org/docs/15/queries-with.html)

### CODE REVIEW

File: pinot-query-runtime/src/test/resources/queries/WithStatements.json [hunk 1]
```diff
@@ -0,0 +1,81 @@
+{
+  "with_statement_tests": {
+    "psql": "7.8.1",
+    "tables": {
+      "tbl1": {
+        "schema": [
+          {"name": "strCol", "type": "STRING"},
```
Comment (61yao) [inline pinot-query-runtime/src/test/resources/queries/WithStatements.json:7]: Can we add select .. from (With as x ( select from tbl)) 

With a as (select .. from tbl) , b as (select .. from a), c as (select ... from  join a,b ) select .. from c.

Comment (walterddr) [reply pinot-query-runtime/src/test/resources/queries/WithStatements.json:7]: could you be more specific?

- `select .. from (With as x ( select from tbl))` is not a valid syntax.
- `with (...) AS a, with (...) AS b SELECT (...) FROM ` is already added. did you meant you would like to nest the previously with tbl in another WITH?

File: pinot-query-runtime/src/test/resources/queries/WithStatements.json [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+          ["a", "bar", 2],
+          ["b", "alice", 42],
+          ["b", "bob", 196883]
+        ]
+      }
+    },
+    "queries": [
+      {
+        "description": "single with with group by",
+        "sql": "WITH w AS ( SELECT strCol, SUM(intCol) AS sumRes FROM {tbl1} GROUP BY strCol ) SELECT strCol, strCol2, sumRes FROM w JOIN {tbl2} ON w.strCol = {tbl2}.strCol1",
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/test/resources/queries/WithStatements.json:34]: Can we add an example where the `JOIN` in outer query is done between `WITH` virtual table W and the same table (`tbl1`) used in `WITH` ?

Comment (siddharthteotia) [reply pinot-query-runtime/src/test/resources/queries/WithStatements.json:34]: Basically WITH can serve for decoration as well (using self-join) where we want to enhance the result of a standard GROUP BY aggregation query on a table with additional columns from the same table.

Something like ....

```
WITH maxResult AS (
  SELECT
    groupCol1,
    groupCol2,
    MAX(aggCol) AS M
  FROM T
  GROUP BY groupCol1, groupCol2
)
SELECT T.*, maxResult.M
 FROM T
 JOIN 
 maxResult
 ON 
 T.groupCol1 = maxResult.groupCol1
 AND 
 T.groupCol2 = maxResult.groupCol2
```

Comment (siddharthteotia) [reply pinot-query-runtime/src/test/resources/queries/WithStatements.json:34]: Can we also add one example for transform inside agg.  ? 

Here is a useful example in the context of a non JOIN query...

```
WITH metricResult AS ( 
      SELECT 
         groupCol1, 
         groupCol2, 
         SUM(c1 * c2) AS metric 
      FROM T 
      GROUP BY groupCol1, groupCol2
) 
SELECT 
  groupCol1,
  MAX(metricResult.metric)
  FROM metricResult
  GROUP BY groupCol1
  ORDER BY MAX(metricResult.metric) DESC
```

File: pinot-query-runtime/src/test/resources/queries/WithStatements.json [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        "outputs": [
+          ["a", "foo", 4],
+          ["a", "bar", 4],
+          ["b", "alice", 2],
+          ["b", "bob", 2]
+        ]
+      },
+      {
+        "description": "multi with table",
+        "sql": "WITH agg1 AS ( SELECT strCol, sum(intCol) AS sumVal FROM {tbl1} GROUP BY strCol), agg2 AS (SELECT strCol1, avg(intCol) AS avgVal FROM {tbl2} GROUP BY strCol1) SELECT strCol, sumVal - avgVal FROM agg1, agg2 WHERE agg1.strCol = agg2.strCol1",
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/test/resources/queries/WithStatements.json:44]: Can we also project other aggregations computed in the respective CTEs in the outer query and also try some other transform ? Something like following

```
SELECT 
strCol
sumVal AS agg1_sum,
avgVal AS agg2_avg,
sumVal + avgVal,
sumVal - avgVal
FROM agg1, agg2
WHERE agg1.strCol = agg2.strCol1
```

Same or different query should be fine.

File: pinot-query-runtime/src/test/resources/queries/WithStatements.json [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      {
+        "description": "with collapse into single query",
+        "sql": "WITH can_collapse AS ( SELECT strCol, intCol FROM {tbl1} ) SELECT * FROM can_collapse WHERE intCol > 2",
+        "outputs": [
+          ["a", 3],
+          ["c", 5]
+        ]
+      },
+      {
+        "description": "with statement table being reused multiple times",
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/test/resources/queries/WithStatements.json:59]: May be we can call this nested CTE ?

File: pinot-query-runtime/src/test/resources/queries/WithStatements.json [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        "sql": "WITH reused AS ( SELECT strCol, intCol FROM {tbl1} ) SELECT a.strCol, a.intCol, b.intCol FROM reused AS a JOIN reused AS b ON a.strCol = b.strCol WHERE b.intCol > ( SELECT AVG(intCol) FROM reused )",
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/test/resources/queries/WithStatements.json:60]: Shall we add test for RECURSIVE CTE as well ?

File: pinot-query-runtime/src/test/resources/queries/WithStatements.json (hunk 1 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/resources/queries/WithStatements.json:7]: this is added

File: pinot-query-runtime/src/test/resources/queries/WithStatements.json (hunk 2 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/resources/queries/WithStatements.json:34]: this is added

File: pinot-query-runtime/src/test/resources/queries/WithStatements.json (hunk 3 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/resources/queries/WithStatements.json:44]: added

File: pinot-query-runtime/src/test/resources/queries/WithStatements.json (hunk 5 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/resources/queries/WithStatements.json:83]: added but it is not working

**Review by siddharthteotia — APPROVED** [review]

**Comment by siddharthteotia** [comment]
Thanks for adding all the tests

**Pull request merged**